database/src/ops/block.rs

```diff
@@ -1,89 +1,321 @@
-pub fn pop_block() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `add_block_*`
+/// Add a [`VerifiedBlockInformation`] to the database.
+///
+/// This extracts all the data from the input block and
+/// maps and adds them to the appropriate database tables.
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*};
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[allow(clippy::too_many_lines)]
+// no inline, too big.
+pub fn add_block(
+    tables: &mut impl TablesMut,
+    block: &VerifiedBlockInformation,
+) -> Result<(), RuntimeError> {
+    // Block Info.
+    //
+    // Branch on the hard fork version (`major_version`)
+    // and add the block to the appropriate table.
+    // <https://monero-book.cuprate.org/consensus_rules/hardforks.html#Mainnet-Hard-Forks>
+    //
+    // FIXME: use `match` with ranges when stable:
+    // <https://github.com/rust-lang/rust/issues/37854>
+    if block.block.header.major_version < 4 {
+        tables.block_info_v1s_mut().put(
+            &block.height,
+            &BlockInfoV1 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                #[allow(clippy::cast_possible_truncation)] // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+            },
+        )
+    } else if block.block.header.major_version < 10 {
+        tables.block_info_v2s_mut().put(
+            &block.height,
+            #[allow(clippy::cast_possible_truncation)] // TODO
+            &BlockInfoV2 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                #[allow(clippy::cast_possible_truncation)] // TODO
+                cumulative_difficulty: block.cumulative_difficulty as u64, // TODO
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(), // TODO
+            },
+        )
+    } else {
+        tables.block_info_v3s_mut().put(
+            &block.height,
+            &BlockInfoV3 {
+                timestamp: block.block.header.timestamp,
+                total_generated_coins: block.generated_coins,
+                weight: block.weight as u64, // TODO
+                cumulative_difficulty: block.cumulative_difficulty,
+                block_hash: block.block_hash,
+                cumulative_rct_outs: todo!(),                    // TODO
+                long_term_weight: block.long_term_weight as u64, // TODO
+            },
+        )
+    }?;
 
-/// TODO
-pub fn block_exists() {
-    todo!()
-}
+    // Block blobs.
+    // TODO: what is a block blob in Cuprate's case?
+    tables.block_blobs_mut().put(&block.height, todo!())?;
```

Comment:
Is this the blob of tx's in the block?
```rust
monero_serai::Block {
    /* ... */
    txs: Vec<[u8; 32]> // <- this?
}
```

Reply:
It is the output from this: https://github.com/serai-dex/serai/blob/93be7a30674ecedfb325b6d09dc22d550d7c13f8/coins/monero/src/block.rs#L110 but it can be added to `VerifiedBlockInformation` as we already serialize the block in the consensus code.